### PR TITLE
Feature: Get location

### PR DESCRIPTION
**Pull request by mixcas**
Theres a new field in Theme Options, the api key was sent to globie email.
Fixed some linting too.

### CODE REVIEW

**Comment by mikeyrayvon** [comment]
i think `geolocation()` should be in `Site.Geolocation`, because later we will need to put the initial location into the `#lat-long` elem (found in `partials/map.php`) and update that displayed latlong as user moves around the map.

**Comment by mikeyrayvon** [comment]
so as there is a bit more to it than just getting the initial location and positioning it, i think it should have its own space. agree or nah?

File: lib/dbip-client.class.php [hunk 1]
```diff
@@ -0,0 +1,84 @@
+<?php
```
Comment (varffs) [inline lib/dbip-client.class.php:1]: could this be installed via Composer?

File: lib/functions-custom.php [hunk 2]
```diff
@@ -1,3 +1,34 @@
...
+require "dbip-client.class.php";
 
-// Custom functions (like special queries, etc)
\ No newline at end of file
+// Custom functions (like special queries, etc)
+
+function get_client_geolocation() {
+  $client_ip = get_client_ip();
+
+  $api_key = IGV_get_option('_igv_site_options', '_igv_dbip_api_key');
```
Comment (varffs) [inline lib/functions-custom.php]: You need to check for this first. I think it causes a critical failure if this value is not set.

Comment (varffs) [reply lib/functions-custom.php]: Yes it def causes a failure of the entire theme

File: src/js/main.js [hunk 3]
```diff
@@ -395,37 +400,84 @@ Site.Map = {
...
+    // Animate recursevly
+    window.requestAnimationFrame(_this.pan.bind(_this));
+
+  },
+
+  // Check if golocation is available and move the map to that equivalent location
+  geolocation: function() {
+    var _this =  this;
+
+    // THIS IS FOR DEVELOPING WHILE WE GET THE REAL API KEY
```
Comment (varffs) [inline src/js/main.js]: why is this in the branch? We have an API key

**Review by varffs — CHANGES_REQUESTED** [review]
Needs conditionals in the PHP. Needs real data in the JS. I agree with @mikeyrayvon on the refactor. Also the experience of the loading and moving to the geolocation thing is bad as the jump feels wrong. Either the map should wait to show until after the move, or the move should be animated and visually related to the geolocation text. (Also that text needs to be from real data too!)

File: lib/dbip-client.class.php (hunk 1 above)
Comment (mixcas) [reply lib/dbip-client.class.php:1]: no ;(

File: lib/functions-custom.php (hunk 2 above)
Comment (mixcas) [reply lib/functions-custom.php]: word

File: src/js/main.js (hunk 3 above)
Comment (mixcas) [reply src/js/main.js]: cuz when i was working we didnt have one…?

Comment (mixcas) [reply src/js/main.js]: actually. the one we had was globie one, free plan. free plan doesnt return lat long. thats why

**Comment by mixcas** [comment]
I agree with the missing conditionals in php and some bits in js. Will fix

About refactoring to it own object. I disagree. IMO [getting geolocation and moving the map to the matching position] is one feature and [geolocation indicator updating with coordinates matching the map position] is a diff feature and should be handled in a diff branch and definitely different object.

Also the [coordinates indicator] should be event driven. so whenever `moveMap` ocurres an `movemap` event is triggered . This is was is cleaner code, better performance and we don't have to worry about setting the initial coordinates in the indicator "manually".

**Comment by varffs** [comment]
Ok fine to spec this feature as just that, but if so I think you need to name your PR a bit tighter or point out in the description the scope of it. Important to clean this asap as it is therefore blocking the rest of the geolocation features

**Comment by mixcas** [comment]
I'm just getting to bed. If someone else could finish this would be the best.

**Review by varffs — APPROVED** [review]

**Comment by mikeyrayvon** [comment]
this puts me at the center of the map. maybe this needs some error handling?

File: lib/functions-custom.php [hunk 4]
```diff
@@ -1,3 +1,37 @@
...
+function get_client_geolocation() {
+  $api_key = IGV_get_option('_igv_site_options', '_igv_dbip_api_key');
+
+  if ($api_key) {
+    $dbip = new DBIP_Client($api_key);
+    $client_ip = get_client_ip();
+
+    return $dbip->Get_Address_Info($client_ip);
+  } else {
+    return '0';
```
Comment (mixcas) [inline lib/functions-custom.php]: Why return string here instead of false?

File: src/js/main.js [hunk 5]
```diff
@@ -395,37 +400,84 @@ Site.Map = {
...
+    // Animate recursevly
+    window.requestAnimationFrame(_this.pan.bind(_this));
+
+  },
+
+  // Check if golocation is available and move the map to that equivalent location
+  geolocation: function() {
+    var _this =  this;
+
+    if (WP.clientGeolocation === '0') {
```
Comment (mixcas) [inline src/js/main.js]: Why string? x2

Comment (mixcas) [reply src/js/main.js]: If we `return false` from the php we can check for `WP.clientGeolocation == false` here

**Review by mixcas — COMMENTED** [review]
Fine except those pseudo boolean strings. What's the reason for this approach?

**Comment by mixcas** [comment]
@mikeyrayvon We need the paid api key to get lat/long otherwise the free one doesnt return those values

File: src/js/main.js (hunk 5 above)
Comment (varffs) [reply src/js/main.js]: I guess I wasn't sure about that due to the PHP>JS thing, but I'll test and change if it works

**Comment by varffs** [comment]
Yh I didn't realise when I did my edits that the current key was only test so assumed it was going to work. Atm the returned values from the API don't include latlong. We could wait to merge this but I think should just do it now and fix any issues if they arrise with the real API. (I assumed @mixcas knew the response was going to have the `latitude` property named as he had used in testing)

**Comment by varffs** [comment]
Well as I thought returning false from the PHP `false` return doesn't exactly translate into the JS. The JS actually ends up returning a empty string, but that reads as false with JS so the final outcome is cleaner code

**Comment by mikeyrayvon** [comment]
Paid API Key:
a82de32adcadd64016832aa0d242ad460fe97b49

**Comment by mikeyrayvon** [comment]
had to sign in to github at lumen to delete that comment. but api key is in trello card lol

File: src/js/main.js [hunk 6]
```diff
@@ -421,37 +426,82 @@ Site.Map = {
...
+
+      // Transport latitude to only negative values so it goes from -180 to 0
+      var translatedLatitude = latitude - 90;
+
+      // Magic math, jk. It'ssic math.
+      var newX = (translatedLongitude * (_this.window.width * -2)) / 360;
+      var newY = (translatedLatitude * (_this.window.height * 2)) / 180;
+
+      // Move the map
+      _this.moveMap(newX,newY);
```
Comment (varffs) [inline src/js/main.js:476]: this function `moveMap()` calls `Site.Minimap.moveIndicator()` but does so before the Minimap object has been initialized. That `moveIndicator()` function fails as it references something cached in the init function which hasn't happened. This breaks all the js. This also relates to what @mixcas was talking about in #17

**Review by varffs — CHANGES_REQUESTED** [review]
I'm not sure the solution. Either refactoring Minimap in a quick way or a complex way. Or a hack for the moment. or?

File: src/js/main.js (hunk 6 above)
Comment (varffs) [reply src/js/main.js:476]: This is the line that causes the issue: https://github.com/interglobalvision/principled-design-com/blob/master/src/js/main.js#L614

Comment (mixcas) [reply src/js/main.js:476]: @varffs Yeah, this is exactly what i am talking about in https://github.com/interglobalvision/principled-design-com/issues/17

This error you are talking about is not happening to me tho. Maybe is a problem with your built files? How about you @mikeyrayvon ?

Comment (mixcas) [reply src/js/main.js:476]: Ah sorry. I didn't pull.

@varffs you fucked it in the merge. I will fix it

Why did u you merge and push anyway?

**Comment by mixcas** [comment]
@varffs i really don't get why u merged and push

File: src/js/main.js [hunk 7]
```diff
@@ -14,8 +14,8 @@ Site = {
 
     $(document).ready(function () {
       Site.Shapes.init();
-      Site.Map.init();
       Site.Minimap.init();
+      Site.Map.init();
```
Comment (varffs) [inline src/js/main.js]: Why did you revert this? Minimap depends on Map and should be init after anyway?

File: src/js/main.js (hunk 6 above)
Comment (varffs) [reply src/js/main.js:476]: Was trying to make it a one button press for someone to speed it up

File: src/js/main.js (hunk 7 above)
Comment (mixcas) [reply src/js/main.js]: Cuz otherwise it wasn't working. I was thinking we should move the `window` object from `Site.Map` into just `Site`

**Pull request merged**